### ISSUE TITLE
jenkins_plugin reports "change = true" even when nothing has changed

### ISSUE THREAD

**Issue by marcdb** [ansible/ansible-modules-extras#3600]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
jenkins_plugin

##### ANSIBLE VERSION
```
ansible 2.2.0.0
  config file = /home/ext950/GIT/marc/ansible.cfg
  configured module search path = Default w/o overrides
```

##### OS / ENVIRONMENT

managing host: Ubuntu 14.04 LTS
managed host: CentOS Linux release 7.2.1511 (Core)  (vagrant box)

##### SUMMARY
Each time I execute a playbook containing a task which calls the jenkins_plugin module, it reports that changes have been made. Even if the requested version of the jenkins plugin was already installed and nothing has changed.

##### STEPS TO REPRODUCE

Below is an extract from my playbook:

```
- name: install jenkins plugins
  jenkins_plugin:
    owner: '{{ jenkins_user }}'
    jenkins_home: '{{ jenkins_home }}'
    name: '{{ item.name }}'
    version: '{{ item.version }}'
    state: pinned
    params:
      url_username: '{{ jenkins_admin_user }}'
      url_password: '{{ jenkins_admin_pass }}'
  register: result
  until: result|success
  retries: 1
  delay: 1
  with_items: '{{ jenkins_plugins }}'
  become: yes
  become_user: '{{ jenkins_user }}'
  notify: restart jenkins
```
The variable jenkins_plugins looks like this:
```
jenkins_plugins:
  - name: postbuild-task
    version: latest
  - name: build-environment
    version: 1.6
```

##### EXPECTED RESULTS
changed should be set to false if the requested version of the plugin is already installed.

##### ACTUAL RESULTS

"changed" is always set to "true". The example is created when requesting the latest version, but also when requesting a fixed version, "changed" is set to "true".

```
changed: [vag-build-blo] => (item={u'version': u'latest', u'name': u'postbuild-task'}) => {
    "attempts": 1, 
    "changed": true, 
    "invocation": {
        "module_args": {
            "backup": null, 
            "content": null, 
            "delimiter": null, 
            "directory_mode": null, 
            "follow": false, 
            "force": false, 
            "force_basic_auth": true, 
            "group": "jenkins", 
            "http_agent": "ansible-httpget", 
            "jenkins_home": "/var/lib/jenkins", 
            "mode": "0644", 
            "name": "postbuild-task", 
            "owner": "jenkins", 
            "regexp": null, 
            "remote_src": null, 
            "selevel": null, 
            "serole": null, 
            "setype": null, 
            "seuser": null, 
            "src": null, 
            "state": "pinned", 
            "timeout": 30.0, 
            "unsafe_writes": null, 
            "updates_expiration": 86400, 
            "updates_url": "https://updates.jenkins-ci.org", 
            "url": "http://localhost:8080", 
            "url_password": "xxxxxxxxxxxxxx", 
            "url_username": "yyyyyyy", 
            "use_proxy": true, 
            "validate_certs": true, 
            "version": "latest", 
            "with_dependencies": true
        }, 
        "module_name": "jenkins_plugin"
    }, 
    "item": {
        "name": "postbuild-task", 
        "version": "latest"
    }, 
    "plugin": "postbuild-task", 
    "state": "pinned"
}
```

**Comment by ansibot** [comment]
@jtyr ping, this issue is waiting for your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jtyr** [comment]
Thanks for your report, @marcdb. The issue here is that Jenkins doesn't know about new plugins until you restart it. This is why the installation returns `changed` again if you run the installation tasks multiple times. Please refer to the examples from the module documentation for more information how to deal with this issue. Marking this as `notabug`.

**Comment by marcdb** [comment]
Hello @jtyr,
Actually, I do restart jenkins after the installation of the plugins. The "notify: restart jenkins" will make sure that a handler is called to restart Jenkins. The problem now is that this handler is called each time I run this playbook. I would expect that the second time I run it, the "change" field will be false. But it isn't in my case. I assume from your reply that in your tests, the change field remains false in such case ?

**Comment by jtyr** [comment]
The notification restarts Jenkins only at the end of the Play, not right after the plugin was installed. Try the following and you will see that it won't show `changed` in the last plugin installation task:

```
- name: Jenkins Master play
  hosts: jenkins-master
  tasks:
    # This should return "changed" because the plugin is not installed
    - name: Install plugin for the first time
      jenkins_plugin:
        name: token-macro

    # This should return "changed" although the plugin was already once installed
    # but the Jenkins server wasn't restarted yet
    - name: Install plugin for the first time
      jenkins_plugin:
        name: token-macro

    - name: Restart Jenkins
      service:
        name: jenkins
        state: restarted

    - name: Wait for Jenkins to start up
      uri:
        url: http://localhost:8080
        status_code: 200
        timeout: 5
      register: jenkins_service_status
      # Keep trying for 5 mins in 5 sec intervals
      retries: 60
      delay: 5
      until: >
         'status' in jenkins_service_status and
         jenkins_service_status['status'] == 200

    # This should NOT return "changed" because the plugin was already installed
    # and the Jenkins server restarted
    - name: Install plugin for the second time
      jenkins_plugin:
        name: token-macro
```

**Comment by marcdb** [comment]
The restart of jenkins in our case, is done through a handler, which we flush after the installation of the plugin. So Jenkins is really restarted. Before I run the play again. So that cannot be the real problem. So, I have investigated this a bit further.

It looks that it is related to the setting "state: pinned". Looking at the code, I have the impression that the plugin must already be installed before it can be pinned. That condition is fulfilled. Then the module tries to pin my plugins. Allthough I don't get any error for this, the plugins were not pinned after having executed the play (verified via the Jenkins console).  The function  _pm_query calls in that case _get_url_data. _get_url_data doesn't return an error, but after having executed this, I can see in the Jenkins console that the plugins are not pinned.

To test this further, I have also changed the requested state to "present".
In that case, I actually get real errors related to my environment:
```
"Failed to validate the SSL certificate for updates.jenkins-ci.org:443. Make sure your managed systems have a valid CA certificate installed. You can use validate_certs=False if you do not need to confirm the servers identity but this is unsafe and not recommended. Paths checked for this platform: /etc/ssl/certs, /etc/pki/ca-trust/extracted/pem, /etc/pki/tls/certs, /usr/share/ca-certificates/cacert.org, /etc/ansible"
```
This is something I have to investigate further. 

I suspect now that the call to get_url will not do what it is supposed to do because of the same certificate problem. This function doesn't return any error however. So each time when I request to set the state to "pinned", ansible will tell me that something was changed, allthough in reality it was not changed.
